compiler/rustc_mir/src/borrow_check/consumers.rs

```diff
@@ -0,0 +1,28 @@
+//! This file provides API for compiler consumers.
+
+use rustc_hir::def_id::LocalDefId;
+use rustc_index::vec::IndexVec;
+use rustc_infer::infer::TyCtxtInferExt;
+use rustc_middle::mir::Body;
+use rustc_middle::ty::{self, TyCtxt};
+
+pub use super::{
+    facts::{AllFacts as PoloniusInput, RustcFacts},
+    location::{LocationTable, RichLocation},
+    nll::PoloniusOutput,
+    BodyWithBorrowckFacts,
+};
+
+/// This function computes Polonius facts for the given body. It makes a copy of
+/// the body because it needs to regenerate the region identifiers.
+pub fn get_body_with_borrowck_facts<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def: ty::WithOptConstParam<LocalDefId>,
+) -> BodyWithBorrowckFacts<'tcx> {
+    let (input_body, promoted) = tcx.mir_promoted(def);
+    tcx.infer_ctxt().enter(|infcx| {
+        let input_body: &Body<'_> = &input_body.borrow();
```

Comment:
You have to ensure that you call this function before the mir is stolen by `mir_borrowck`. I don't think you can easily do this. If the mir is used during const-eval, `mir_borrowck` will be called on it. Const-eval can happen during typeck.

Reply:
Prusti passes all tests with this change, so it does not seem to be a problem at the moment (at least for the examples we currently support). However, at some point, it used to be a problem and then we solved it by overriding the [`mir_borrowck`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/query/struct.Providers.html#structfield.mir_borrowck) query in [`Config`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/interface/struct.Config.html#structfield.override_queries). Do you see any reason why our backup approach could fail?

Reply:
Overriding `mir_borrowck` should work.

Does prusti work fine with this PR for

```rust
const fn foo() -> u8 {
    1
}

fn main() {
    let bar: [u8; foo()] = [1];
    assert_eq!(bar[0], foo());
}
```

where `foo` is used both at compiletime and runtime?

Reply:
Thanks for the example! Prusti crashes with both the old and the new APIs (in both cases the error is `attempted to read from stolen value`). So, it seems that I will have to implement the query overriding.

Reply:
I have [updated Prusti](https://github.com/viperproject/prusti-dev/blob/0e718e6334f31a0f60f4489d84828b91d97c7ab5/prusti/src/callbacks.rs#L18-L39) to use the overridden `mir_borrowck` query. This fixed the crash. One downside of this approach is that it required me to use unsafe to [transmute `'tcx` to `'static`](https://github.com/viperproject/prusti-dev/blob/0e718e6334f31a0f60f4489d84828b91d97c7ab5/prusti-interface/src/environment/mir_storage.rs#L28-L31) so that I could store `mir::Body<'tcx>` in a thread-local variable. In case you have any ideas on how to avoid using unsafe in this case, I would be more than happy to hear.

Reply:
I don't think you can avoid the transmute.

Reply:
We should document this requirement somewhere.

Reply:
I have added a comment on the function. Or do you mean to document it somewhere more prominently?